pkg/operation/botanist/component/extensions/operatingsystemconfig/operatingsystemconfig.go

```diff
@@ -693,14 +693,22 @@ func (d *deployer) deploy(ctx context.Context, operation string) (extensionsv1al
 	return d.osc, err
 }
 
-// Key returns the key that can be used as secret name based on the provided worker name and Kubernetes version.
-func Key(workerName string, kubernetesVersion *semver.Version) string {
+// Key returns the key that can be used as secret name based on the provided worker name, Kubernetes version and CRI configuration.
+func Key(workerName string, kubernetesVersion *semver.Version, criConfig *gardencorev1beta1.CRI) string {
 	if kubernetesVersion == nil {
 		return ""
 	}
 
-	kubernetesMajorMinorVersion := fmt.Sprintf("%d.%d", kubernetesVersion.Major(), kubernetesVersion.Minor())
-	return fmt.Sprintf("cloud-config-%s-%s", workerName, utils.ComputeSHA256Hex([]byte(kubernetesMajorMinorVersion))[:5])
+	var (
+		kubernetesMajorMinorVersion = fmt.Sprintf("%d.%d", kubernetesVersion.Major(), kubernetesVersion.Minor())
+		criName                     gardencorev1beta1.CRIName
+	)
+
+	if criConfig != nil && criConfig.Name != gardencorev1beta1.CRINameDocker {
```

Comment:
Wouldn't it make sense to change that to?
```go 
	if criConfig != nil {
		criName = criConfig.Name
	}
```
By doing so we could still add the legacy configuration and for the future docker and containerd secrets would share the same secret name scheme.

Reply:
The goal of checking `cri.Name != docker` is to avoid any changes happening when people switch from the implicit configuration of `docker` (having `cri: nil`) to the explicit configuration (by specifying `cri.name: docker`. This isn't really changing anything on the nodes, therefore we don't want any node rollout or other disruptive stuff happening.
You're right that this is still the case with your proposed change – as long as the compatibility code in `https://github.com/gardener/gardener/pull/5870/files/3f584e4380ba1fecca393194a6c3329047413fa4#diff-78a67ae80f891c299fbd162c192ef1fe19663c5db0bef7e1a42bb04eb70a6335` is present. Once we remove this compatibility code, nodes will no longer find their old secret when configuring `docker` explicitly moves to a new secret name.
We could achieve a similar thing with a slightly different change, though:
```
if criConfig == nil {
		criName = gardencorev1beta1.CRINameDocker
	}
```

This would move all `cri: nil` clusters to the new cloud-config secret and after removing the compatibility code, switching from `cri: nil` to `cri.name: docker` still works without any sideeffects.

Reply:
Avoiding unneeded rollouts when changing from docker to nil and vice versa definitely makes sense.

For my understanding, I'm pretty new to this stuff 😄 
- the current approach would roll every containerd node as a side effect when this PR is applied, because the secret name is changing in this case. Docker nodes (implicit and explicit) would not roll.
-  In your proposal every node would roll initially, because all secret names are changing

Is this correct?

Reply:
We're not rolling nodes as an effect of this PR – nodes would be rolled when people update the CRI configuration in their worker pools. This change is aimed at removing an effect we saw when this node rollout happens: the "old nodes" still apply a cloud-config that wasn't meant for them, because it contains the new configuration.

What the compatibility part of this change does is that it adds a second cloud-config secret for existing nodes which are already configured with `cri.name: containerd` – to avoid [the problem we experienced when we tried introducing cloud-configs with names based on the CRI](https://github.com/gardener/gardener/issues/4390): you cannot simply change the name of a cloud-config for a running node, because each node runs a script constantly checking the cloud-config secret for new content (you can see how these components interact [in a diagram included in the bootstrap documentation](https://github.com/gardener/gardener/blob/master/docs/extensions/operatingsystemconfig.md#detailed-bootstrap-flow-with-a-worker-generated-bootstrap-token)). There's a tricky self-updating mechanism in place here where the downloaded script can contain an updated version of the updater script, such that in the next run, the new secret name will be checked for updated content.

What I was proposing above changes @rfranzke's approach slightly in the sense that we not only do this for existing nodes configured with `containerd`, but also for the ones with `docker`, such that the computation of the cloud-config-secret name will not contain any "special if-else magic" for particular runtimes, but just computes a hash. I guess this is also where your simplification proposal on top of this thread came from?

Reply:
I would propose to proceed as is and live with this "special magic". I haven't checked, but we probably have other places where we assume that `cri=nil` is the same as `cri.name=docker`. "Simplifying" this now would require me to re-validate this PR for all the various cases and I would rather spend the time on something else.